File: 122-best-time-to-buy-and-sell-stock-ii/best-time-to-buy-and-sell-stock-ii.java
class Solution {
    public int maxProfit(int[] nums) {
        // return profit(nums,1,0);
        int[][] dp=new int[nums.length][2];
        for(int[]a:dp){
            Arrays.fill(a,-1);
        }
        return TD(nums,1,0,dp);
    }

    public int profit(int[] arr,int buy,int idx){
        if(idx>=arr.length){
            return 0;
        }
        int spend=0;
        int sell=0;
        if(buy==1){
            spend=Math.max(-arr[idx]+profit(arr,0,idx+1),profit(arr,1,idx+1));
        }
        else{
            sell=Math.max(arr[idx]+profit(arr,1,idx+1),profit(arr,0,idx+1));
        }
        return Math.max(sell,spend);
    }
    public int TD(int[] arr,int buy,int idx,int[][] dp){
        if(idx>=arr.length){
            return 0;
        }
        if(dp[idx][buy]!=-1){
            return dp[idx][buy];
        }
        int spend=0;
        int sell=0;
        if(buy==1){
            spend=Math.max(-arr[idx]+TD(arr,0,idx+1,dp),TD(arr,1,idx+1,dp));
        }
        else{
            sell=Math.max(arr[idx]+TD(arr,1,idx+1,dp),TD(arr,0,idx+1,dp));
        }
        return dp[idx][buy]=Math.max(sell,spend);
    }

}

File: 122-best-time-to-buy-and-sell-stock-ii/best-time-to-buy-and-sell-stock-ii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        // return profit(0,prices,1);
        int ans=0;
        vector<vector<int>> dp(prices.size()+1,vector<int>(2,0));
        for(int i=prices.size()-1;i>=0;i--){
            for(int j=0;j<2;j++){
                if(j==1){
                    dp[i][j]=max(-prices[i]+dp[i+1][0],dp[i+1][1]);
                }
                else{
                    dp[i][j]=max(prices[i]+dp[i+1][1],dp[i+1][0]);
                }
            }
        }
        return dp[0][1];
    }
    int profit(int idx,vector<int> &arr,int turn){
        if(idx==arr.size()){
            return 0;
        }

        int buy=0;
        int sell=0;
        if(turn){
            buy=max(-arr[idx]+profit(idx+1,arr,0),profit(idx+1,arr,1));
        }
        else{
            sell=max(arr[idx]+profit(idx+1,arr,1),profit(idx+1,arr,0));
        }
        return max(buy,sell);
    }
};